src/qhull_wrap.cpp

```diff
@@ -168,107 +193,71 @@ delaunay_impl(int npoints, const double* x, const double* y,
-    }
+    /* Allocate Python arrays to return. */
+    npy_intp dims[2] = {ntri, 3};
+    numpy::array_view<int, ndim> triangles(dims);
+    int* triangles_ptr = triangles.data();
 
-    neighbors = (PyArrayObject*)PyArray_SimpleNew(ndim, dims, NPY_INT);
-    if (neighbors == NULL) {
-        PyErr_SetString(PyExc_MemoryError,
-                        "Could not allocate neighbors array in qhull.delaunay");
-        goto error;
-    }
-
-    triangles_ptr = (int*)PyArray_DATA(triangles);
-    neighbors_ptr = (int*)PyArray_DATA(neighbors);
+    numpy::array_view<int, ndim> neighbors(dims);
+    int* neighbors_ptr = neighbors.data();
 
     /* Determine triangles array and set tri_indices array. */
     i = 0;
     FORALLfacets {
         if (!facet->upperdelaunay) {
+            int indices[3];
             tri_indices[facet->id] = i++;
             get_facet_vertices(qh, facet, indices);
             *triangles_ptr++ = (facet->toporient ? indices[0] : indices[2]);
             *triangles_ptr++ = indices[1];
             *triangles_ptr++ = (facet->toporient ? indices[2] : indices[0]);
         }
-        else
+        else {
             tri_indices[facet->id] = -1;
+        }
     }
 
     /* Determine neighbors array. */
     FORALLfacets {
         if (!facet->upperdelaunay) {
+            int indices[3];
             get_facet_neighbours(facet, tri_indices, indices);
             *neighbors_ptr++ = (facet->toporient ? indices[2] : indices[0]);
             *neighbors_ptr++ = (facet->toporient ? indices[0] : indices[2]);
             *neighbors_ptr++ = indices[1];
         }
     }
 
-    /* Clean up. */
-    qh_freeqhull(qh, !qh_ALL);
-    qh_memfreeshort(qh, &curlong, &totlong);
-    if (curlong || totlong)
-        PyErr_WarnEx(PyExc_RuntimeWarning,
-                     "Qhull could not free all allocated memory", 1);
-    if (hide_qhull_errors)
-        fclose(error_file);
-    free(tri_indices);
-    free(points);
-
-    tuple = PyTuple_New(2);
-    PyTuple_SetItem(tuple, 0, (PyObject*)triangles);
-    PyTuple_SetItem(tuple, 1, (PyObject*)neighbors);
-    return tuple;
+    PyObject* tuple = PyTuple_New(2);
+    if (tuple == 0) {
+        throw std::runtime_error("Failed to create Python tuple");
+    }
 
-error:
-    /* Clean up. */
-    Py_XDECREF(triangles);
-    Py_XDECREF(neighbors);
-    qh_freeqhull(qh, !qh_ALL);
-    qh_memfreeshort(qh, &curlong, &totlong);
-    /* Don't bother checking curlong and totlong as raising error anyway. */
-    if (hide_qhull_errors)
-        fclose(error_file);
-    free(tri_indices);
-
-error_before_qhull:
-    free(points);
-
-    return NULL;
+    PyTuple_SET_ITEM(tuple, 0, triangles.pyobj());
+    PyTuple_SET_ITEM(tuple, 1, neighbors.pyobj());
+    return tuple;
 }
 
-/* Process python arguments and call Delaunay implementation method. */
+/* Process Python arguments and call Delaunay implementation method. */
 static PyObject*
 delaunay(PyObject *self, PyObject *args)
 {
     PyObject* xarg;
```

Comment:
You could use the `array_view` converters to avoid having to have any `PyObject*` in this function.  Something like
https://github.com/QuLogic/matplotlib/blob/c62ca9a54d50b91ca971af955fe0f3760bd5deae/src/tri/_tri_wrapper.cpp#L33-L43

Reply:
Right, I forgot to check if those existed.